File: frontend/src/models/sitemaps.ts
import { dto } from "@/wailsjs/wailsjs/go/models";

// =========================================================================
// Sitemap Types
// =========================================================================

export type SitemapSource = "manual" | "imported" | "generated" | "scanned";
export type SitemapStatus = "draft" | "active" | "archived";
export type NodeSource = "manual" | "imported" | "generated" | "scanned";
export type NodeContentType = "page" | "post" | "none";

export type NodeDesignStatus = "draft" | "ready" | "approved";
export type NodeGenerationStatus = "none" | "queued" | "generating" | "generated" | "failed";
export type NodePublishStatus = "none" | "publishing" | "draft" | "pending" | "published" | "failed";

export interface Sitemap {
    id: number;
    siteId: number;
    name: string;
    description?: string;
    source: SitemapSource;
    status: SitemapStatus;
    createdAt: string;
    updatedAt: string;
}

export interface SitemapNode {
    id: number;
    sitemapId: number;
    parentId?: number;
    title: string;
    slug: string;
    description?: string;
    isRoot: boolean;
    depth: number;
    position: number;
    path: string;
    contentType: NodeContentType;
    articleId?: number;
    wpPageId?: number;
    wpUrl?: string;
    source: NodeSource;
    isSynced: boolean;
    lastSyncedAt?: string;
    wpTitle?: string;
    wpSlug?: string;
    isModified: boolean;
    designStatus: NodeDesignStatus;
    generationStatus: NodeGenerationStatus;
    publishStatus: NodePublishStatus;
    isModifiedLocally: boolean;
    lastError?: string;
    positionX?: number;
    positionY?: number;
    keywords: string[];
    children?: SitemapNode[];
    createdAt: string;
    updatedAt: string;
}

export interface SitemapWithNodes {
    sitemap: Sitemap;
    nodes: SitemapNode[];
}

// =========================================================================
// Input Types
// =========================================================================

export interface CreateSitemapInput {
    siteId: number;
    name: string;
    description?: string;
    source: SitemapSource;
    siteUrl: string;
}

export interface UpdateSitemapInput {
    id: number;
    name: string;
    description?: string;
    status: SitemapStatus;
}

export interface CreateNodeInput {
    sitemapId: number;
    parentId?: number;
    title: string;
    slug: string;
    description?: string;
    position?: number;
    source?: NodeSource;
    keywords?: string[];
}

export interface UpdateNodeInput {
    id: number;
    title: string;
    slug: string;
    description?: string;
    keywords?: string[];
}

export interface MoveNodeInput {
    nodeId: number;
    newParentId?: number;
    position?: number;
}

export interface UpdateNodePositionsInput {
    nodeId: number;
    positionX: number;
    positionY: number;
}

export interface ImportNodesInput {
    sitemapId: number;
    parentNodeId?: number;
    filename: string;
    fileDataBase64: string;
}

export interface ImportError {
    row?: number;
    column?: string;
    message: string;
}

export interface ImportNodesResult {
    totalRows: number;
    nodesCreated: number;
    nodesSkipped: number;
    errors: ImportError[];
    processingTime: string;
}

// =========================================================================
// Mappers
// =========================================================================

export function mapSitemap(x: dto.Sitemap): Sitemap {
    return {
        id: x.id,
        siteId: x.siteId,
        name: x.name,
        description: x.description || undefined,
        source: x.source as SitemapSource,
        status: x.status as SitemapStatus,
        createdAt: x.createdAt,
        updatedAt: x.updatedAt,
    };
}

export function mapSitemapNode(x: dto.SitemapNode): SitemapNode {
    return {
        id: x.id,
        sitemapId: x.sitemapId,
        parentId: x.parentId || undefined,
        title: x.title,
        slug: x.slug,
        description: x.description || undefined,
        isRoot: x.isRoot || false,
        depth: x.depth,
        position: x.position,
        path: x.path,
        contentType: x.contentType as NodeContentType,
        articleId: x.articleId || undefined,
        wpPageId: x.wpPageId || undefined,
        wpUrl: x.wpUrl || undefined,
        source: x.source as NodeSource,
        isSynced: x.isSynced,
        lastSyncedAt: x.lastSyncedAt || undefined,
        wpTitle: x.wpTitle || undefined,
        wpSlug: x.wpSlug || undefined,
        isModified: x.isModified || false,
        designStatus: x.designStatus as NodeDesignStatus,
        generationStatus: x.generationStatus as NodeGenerationStatus,
        publishStatus: x.publishStatus as NodePublishStatus,
        isModifiedLocally: x.isModifiedLocally || false,
        lastError: x.lastError || undefined,
        positionX: x.positionX || undefined,
        positionY: x.positionY || undefined,
        keywords: x.keywords || [],
        children: x.children?.map(mapSitemapNode) || undefined,
        createdAt: x.createdAt,
        updatedAt: x.updatedAt,
    };
}

export function mapSitemapWithNodes(x: dto.SitemapWithNodes): SitemapWithNodes {
    return {
        sitemap: mapSitemap(x.sitemap!),
        nodes: (x.nodes || []).map(mapSitemapNode),
    };
}

export function mapImportNodesResult(x: dto.ImportNodesResponse): ImportNodesResult {
    return {
        totalRows: x.totalRows,
        nodesCreated: x.nodesCreated,
        nodesSkipped: x.nodesSkipped,
        errors: (x.errors || []).map((e) => ({
            row: e.row,
            column: e.column,
            message: e.message,
        })),
        processingTime: x.processingTime,
    };
}

// =========================================================================
// Scanner Types
// =========================================================================

export type TitleSource = "title" | "h1";
export type ContentFilter = "all" | "pages" | "posts";

export interface ScanSiteInput {
    siteId: number;
    sitemapName: string;
    titleSource: TitleSource;
    contentFilter: ContentFilter;
    includeDrafts: boolean;
    maxDepth: number;
}

export interface ScanIntoSitemapInput {
    sitemapId: number;
    parentNodeId?: number;
    titleSource: TitleSource;
    contentFilter: ContentFilter;
    includeDrafts: boolean;
    maxDepth: number;
}

export interface ScanError {
    wpId?: number;
    type?: string;
    title?: string;
    message: string;
}

export interface ScanSiteResult {
    sitemapId: number;
    pagesScanned: number;
    postsScanned: number;
    nodesCreated: number;
    nodesSkipped: number;
    totalDuration: string;
    errors: ScanError[];
}

export function mapScanSiteResult(x: dto.ScanSiteResponse): ScanSiteResult {
    return {
        sitemapId: x.sitemapId,
        pagesScanned: x.pagesScanned,
        postsScanned: x.postsScanned,
        nodesCreated: x.nodesCreated,
        nodesSkipped: x.nodesSkipped,
        totalDuration: x.totalDuration,
        errors: (x.errors || []).map((e) => ({
            wpId: e.wpId,
            type: e.type,
            title: e.title,
            message: e.message,
        })),
    };
}

// =========================================================================
// Sync Types
// =========================================================================

export interface SyncNodesInput {
    siteId: number;
    nodeIds: number[];
}

export interface SyncNodeResult {
    nodeId: number;
    success: boolean;
    error?: string;
}

export interface SyncNodesResult {
    results: SyncNodeResult[];
}

export function mapSyncNodesResult(x: dto.SyncNodesResponse): SyncNodesResult {
    return {
        results: (x.results || []).map((r) => ({
            nodeId: r.nodeId,
            success: r.success,
            error: r.error || undefined,
        })),
    };
}

export interface ChangePublishStatusInput {
    siteId: number;
    nodeId: number;
    newStatus: NodePublishStatus;
}

// =========================================================================
// AI Generation Types
// =========================================================================

export interface TitleInput {
    title: string;
    keywords?: string[];
}

export interface GenerateSitemapStructureInput {
    sitemapId?: number;
    siteId?: number;
    name?: string;
    promptId: number;
    placeholders?: Record<string, string>;
    titles: TitleInput[];
    parentNodeIds?: number[];
    maxDepth?: number;
    includeExistingTree?: boolean;
    providerId: number;
}

export interface GenerateSitemapStructureResult {
    sitemapId: number;
    nodesCreated: number;
    durationMs: number;
}

export function mapGenerateSitemapStructureResult(
    x: dto.GenerateSitemapStructureResponse
): GenerateSitemapStructureResult {
    return {
        sitemapId: x.sitemapId,
        nodesCreated: x.nodesCreated,
        durationMs: x.durationMs,
    };
}

// =========================================================================
// History Types
// =========================================================================

export interface HistoryState {
    canUndo: boolean;
    canRedo: boolean;
    undoCount: number;
    redoCount: number;
    lastAction?: string;
    actionApplied?: string;
}

export function mapHistoryState(x: dto.HistoryState): HistoryState {
    return {
        canUndo: x.canUndo,
        canRedo: x.canRedo,
        undoCount: x.undoCount,
        redoCount: x.redoCount,
        lastAction: x.lastAction,
        actionApplied: x.actionApplied,
    };
}

// =========================================================================
// Page Generation Types
// =========================================================================

export type PageGenerationTaskStatus = "pending" | "running" | "paused" | "completed" | "failed" | "cancelled";
export type PageGenerationNodeStatus = "pending" | "generating" | "publishing" | "completed" | "failed" | "skipped";
export type PublishAs = "draft" | "pending" | "publish";

export type WritingStyle = "professional" | "casual" | "formal" | "friendly" | "technical";
export type ContentTone = "informative" | "persuasive" | "educational" | "engaging" | "authoritative";
export type AutoLinkMode = "none" | "before" | "after";
export type LinkingPhase = "none" | "suggesting" | "applying" | "completed";

export interface ContentSettings {
    wordCount: string; // e.g. "1000" or "800-1200"
    writingStyle: WritingStyle;
    contentTone: ContentTone;
    customInstructions?: string;
    useWebSearch?: boolean; // Enable web search for AI generation
    includeLinks?: boolean; // Include approved links from linking plan
    autoLinkMode?: AutoLinkMode; // Automatic link suggestion mode
    autoLinkProviderId?: number; // Provider for link suggestion (defaults to content provider)
    autoLinkSuggestPromptId?: number; // Prompt for link suggestion (link_suggest category)
    autoLinkApplyPromptId?: number; // Prompt for link insertion (link_apply category)
    maxIncomingLinks?: number; // Max incoming links per page (0 = no limit)
    maxOutgoingLinks?: number; // Max outgoing links per page (0 = no limit)
}

export interface StartPageGenerationInput {
    sitemapId: number;
    nodeIds?: number[];
    providerId: number;
    promptId?: number;
    publishAs: PublishAs;
    placeholders?: Record<string, string>;
    maxConcurrency?: number;
    contentSettings?: ContentSettings;
}

export interface GenerationNodeInfo {
    nodeId: number;
    title: string;
    path: string;
    status: PageGenerationNodeStatus;
    articleId?: number;
    wpPageId?: number;
    wpUrl?: string;
    error?: string;
    startedAt?: string;
    completedAt?: string;
}

export interface GenerationTask {
    id: string;
    sitemapId: number;
    siteId: number;
    totalNodes: number;
    processedNodes: number;
    failedNodes: number;
    skippedNodes: number;
    status: PageGenerationTaskStatus;
    startedAt: string;
    completedAt?: string;
    error?: string;
    nodes?: GenerationNodeInfo[];
    // Linking phase tracking
    linkingPhase?: LinkingPhase;
    linksCreated?: number;
    linksApplied?: number;
    linksFailed?: number;
}

export interface DefaultPrompt {
    name: string;
    systemPrompt: string;
    userPrompt: string;
    placeholders: string[];
}

export function mapGenerationTask(x: dto.GenerationTaskResponse): GenerationTask {
    return {
        id: x.id,
        sitemapId: x.sitemapId,
        siteId: x.siteId,
        totalNodes: x.totalNodes,
        processedNodes: x.processedNodes,
        failedNodes: x.failedNodes,
        skippedNodes: x.skippedNodes,
        status: x.status as PageGenerationTaskStatus,
        startedAt: x.startedAt,
        completedAt: x.completedAt || undefined,
        error: x.error || undefined,
        nodes: x.nodes?.map((n) => ({
            nodeId: n.nodeId,
            title: n.title,
            path: n.path,
            status: n.status as PageGenerationNodeStatus,
            articleId: n.articleId || undefined,
            wpPageId: n.wpPageId || undefined,
            wpUrl: n.wpUrl || undefined,
            error: n.error || undefined,
            startedAt: n.startedAt || undefined,
            completedAt: n.completedAt || undefined,
        })),
        // Linking phase tracking
        linkingPhase: (x.linkingPhase as LinkingPhase) || undefined,
        linksCreated: x.linksCreated || undefined,
        linksApplied: x.linksApplied || undefined,
        linksFailed: x.linksFailed || undefined,
    };
}

export function mapDefaultPrompt(x: dto.DefaultPromptResponse): DefaultPrompt {
    return {
        name: x.name,
        systemPrompt: x.systemPrompt,
        userPrompt: x.userPrompt,
        placeholders: x.placeholders || [],
    };
}

File: internal/dto/sitemaps.go
package dto

import (
	"github.com/davidmovas/postulator/internal/domain/entities"
)

// =========================================================================
// Sitemap DTO
// =========================================================================

type Sitemap struct {
	ID          int64   `json:"id"`
	SiteID      int64   `json:"siteId"`
	Name        string  `json:"name"`
	Description *string `json:"description,omitempty"`
	Source      string  `json:"source"`
	Status      string  `json:"status"`
	CreatedAt   string  `json:"createdAt"`
	UpdatedAt   string  `json:"updatedAt"`
}

func NewSitemap(entity *entities.Sitemap) *Sitemap {
	s := &Sitemap{}
	return s.FromEntity(entity)
}

func (d *Sitemap) ToEntity() *entities.Sitemap {
	createdAt, _ := StringToTime(d.CreatedAt)
	updatedAt, _ := StringToTime(d.UpdatedAt)

	return &entities.Sitemap{
		ID:          d.ID,
		SiteID:      d.SiteID,
		Name:        d.Name,
		Description: d.Description,
		Source:      entities.SitemapSource(d.Source),
		Status:      entities.SitemapStatus(d.Status),
		CreatedAt:   createdAt,
		UpdatedAt:   updatedAt,
	}
}

func (d *Sitemap) FromEntity(entity *entities.Sitemap) *Sitemap {
	d.ID = entity.ID
	d.SiteID = entity.SiteID
	d.Name = entity.Name
	d.Description = entity.Description
	d.Source = string(entity.Source)
	d.Status = string(entity.Status)
	d.CreatedAt = TimeToString(entity.CreatedAt)
	d.UpdatedAt = TimeToString(entity.UpdatedAt)
	return d
}

// =========================================================================
// SitemapNode DTO
// =========================================================================

type SitemapNode struct {
	ID                int64          `json:"id"`
	SitemapID         int64          `json:"sitemapId"`
	ParentID          *int64         `json:"parentId,omitempty"`
	Title             string         `json:"title"`
	Slug              string         `json:"slug"`
	Description       *string        `json:"description,omitempty"`
	IsRoot            bool           `json:"isRoot"`
	Depth             int            `json:"depth"`
	Position          int            `json:"position"`
	Path              string         `json:"path"`
	ContentType       string         `json:"contentType"`
	ArticleID         *int64         `json:"articleId,omitempty"`
	WPPageID          *int           `json:"wpPageId,omitempty"`
	WPURL             *string        `json:"wpUrl,omitempty"`
	Source            string         `json:"source"`
	IsSynced          bool           `json:"isSynced"`
	LastSyncedAt      *string        `json:"lastSyncedAt,omitempty"`
	WPTitle           *string        `json:"wpTitle,omitempty"`
	WPSlug            *string        `json:"wpSlug,omitempty"`
	IsModified        bool           `json:"isModified"`
	DesignStatus      string         `json:"designStatus"`
	GenerationStatus  string         `json:"generationStatus"`
	PublishStatus     string         `json:"publishStatus"`
	IsModifiedLocally bool           `json:"isModifiedLocally"`
	LastError         *string        `json:"lastError,omitempty"`
	PositionX         *float64       `json:"positionX,omitempty"`
	PositionY         *float64       `json:"positionY,omitempty"`
	Keywords          []string       `json:"keywords,omitempty"`
	Children          []*SitemapNode `json:"children,omitempty"`
	CreatedAt         string         `json:"createdAt"`
	UpdatedAt         string         `json:"updatedAt"`
}

func NewSitemapNode(entity *entities.SitemapNode) *SitemapNode {
	n := &SitemapNode{}
	return n.FromEntity(entity)
}

func (d *SitemapNode) ToEntity() *entities.SitemapNode {
	createdAt, _ := StringToTime(d.CreatedAt)
	updatedAt, _ := StringToTime(d.UpdatedAt)

	node := &entities.SitemapNode{
		ID:                d.ID,
		SitemapID:         d.SitemapID,
		ParentID:          d.ParentID,
		Title:             d.Title,
		Slug:              d.Slug,
		Description:       d.Description,
		Depth:             d.Depth,
		Position:          d.Position,
		Path:              d.Path,
		ContentType:       entities.NodeContentType(d.ContentType),
		ArticleID:         d.ArticleID,
		WPPageID:          d.WPPageID,
		WPURL:             d.WPURL,
		Source:            entities.NodeSource(d.Source),
		IsSynced:          d.IsSynced,
		DesignStatus:      entities.NodeDesignStatus(d.DesignStatus),
		GenerationStatus:  entities.NodeGenerationStatus(d.GenerationStatus),
		PublishStatus:     entities.NodePublishStatus(d.PublishStatus),
		IsModifiedLocally: d.IsModifiedLocally,
		LastError:         d.LastError,
		PositionX:         d.PositionX,
		PositionY:         d.PositionY,
		Keywords:          d.Keywords,
		CreatedAt:         createdAt,
		UpdatedAt:         updatedAt,
	}

	if d.LastSyncedAt != nil {
		lastSyncedAt, _ := StringToTime(*d.LastSyncedAt)
		node.LastSyncedAt = &lastSyncedAt
	}

	return node
}

func (d *SitemapNode) FromEntity(entity *entities.SitemapNode) *SitemapNode {
	d.ID = entity.ID
	d.SitemapID = entity.SitemapID
	d.ParentID = entity.ParentID
	d.Title = entity.Title
	d.Slug = entity.Slug
	d.Description = entity.Description
	d.IsRoot = entity.IsRoot
	d.Depth = entity.Depth
	d.Position = entity.Position
	d.Path = entity.Path
	d.ContentType = string(entity.ContentType)
	d.ArticleID = entity.ArticleID
	d.WPPageID = entity.WPPageID
	d.WPURL = entity.WPURL
	d.Source = string(entity.Source)
	d.IsSynced = entity.IsSynced
	d.WPTitle = entity.WPTitle
	d.WPSlug = entity.WPSlug
	d.IsModified = entity.IsModified()
	d.DesignStatus = string(entity.DesignStatus)
	d.GenerationStatus = string(entity.GenerationStatus)
	d.PublishStatus = string(entity.PublishStatus)
	d.IsModifiedLocally = entity.IsModifiedLocally
	d.LastError = entity.LastError
	d.PositionX = entity.PositionX
	d.PositionY = entity.PositionY
	d.Keywords = entity.Keywords
	d.CreatedAt = TimeToString(entity.CreatedAt)
	d.UpdatedAt = TimeToString(entity.UpdatedAt)

	if entity.LastSyncedAt != nil {
		lastSyncedAt := TimeToString(*entity.LastSyncedAt)
		d.LastSyncedAt = &lastSyncedAt
	}

	// Convert children recursively
	if len(entity.Children) > 0 {
		d.Children = make([]*SitemapNode, len(entity.Children))
		for i, child := range entity.Children {
			d.Children[i] = NewSitemapNode(child)
		}
	}

	return d
}

// =========================================================================
// Request/Response DTOs
// =========================================================================

type CreateSitemapRequest struct {
	SiteID      int64   `json:"siteId"`
	Name        string  `json:"name"`
	Description *string `json:"description,omitempty"`
	Source      string  `json:"source"`
	SiteURL     string  `json:"siteUrl"` // Used to create root node
}

type UpdateSitemapRequest struct {
	ID          int64   `json:"id"`
	Name        string  `json:"name"`
	Description *string `json:"description,omitempty"`
	Status      string  `json:"status"`
}

type DuplicateSitemapRequest struct {
	ID      int64  `json:"id"`
	NewName string `json:"newName"`
}

type CreateNodeRequest struct {
	SitemapID   int64    `json:"sitemapId"`
	ParentID    *int64   `json:"parentId,omitempty"`
	Title       string   `json:"title"`
	Slug        string   `json:"slug"`
	Description *string  `json:"description,omitempty"`
	Position    int      `json:"position"`
	Source      string   `json:"source"`
	Keywords    []string `json:"keywords,omitempty"`
}

type UpdateNodeRequest struct {
	ID          int64    `json:"id"`
	Title       string   `json:"title"`
	Slug        string   `json:"slug"`
	Description *string  `json:"description,omitempty"`
	Keywords    []string `json:"keywords,omitempty"`
}

type MoveNodeRequest struct {
	NodeID      int64  `json:"nodeId"`
	NewParentID *int64 `json:"newParentId,omitempty"`
	Position    int    `json:"position"`
}

type UpdateNodePositionsRequest struct {
	NodeID    int64   `json:"nodeId"`
	PositionX float64 `json:"positionX"`
	PositionY float64 `json:"positionY"`
}

type LinkNodeToArticleRequest struct {
	NodeID    int64 `json:"nodeId"`
	ArticleID int64 `json:"articleId"`
}

type LinkNodeToPageRequest struct {
	NodeID   int64  `json:"nodeId"`
	WPPageID int    `json:"wpPageId"`
	WPURL    string `json:"wpUrl"`
}

type SetNodeKeywordsRequest struct {
	NodeID   int64    `json:"nodeId"`
	Keywords []string `json:"keywords"`
}

type DistributeKeywordsRequest struct {
	SitemapID int64    `json:"sitemapId"`
	Keywords  []string `json:"keywords"`
	Strategy  string   `json:"strategy"`
}

type SitemapWithNodes struct {
	Sitemap *Sitemap       `json:"sitemap"`
	Nodes   []*SitemapNode `json:"nodes"`
}

// =========================================================================
// Import DTOs
// =========================================================================

type ImportNodesRequest struct {
	SitemapID      int64  `json:"sitemapId"`
	ParentNodeID   *int64 `json:"parentNodeId,omitempty"` // If set, import as children of this node
	Filename       string `json:"filename"`
	FileDataBase64 string `json:"fileDataBase64"` // Base64 encoded file content
}

type ImportNodesResponse struct {
	TotalRows      int           `json:"totalRows"`
	NodesCreated   int           `json:"nodesCreated"`
	NodesSkipped   int           `json:"nodesSkipped"`
	Errors         []ImportError `json:"errors,omitempty"`
	ProcessingTime string        `json:"processingTime"`
}

type ImportError struct {
	Row     int    `json:"row,omitempty"`
	Column  string `json:"column,omitempty"`
	Message string `json:"message"`
}

type SupportedFormatsResponse struct {
	Formats []string `json:"formats"`
}

// =========================================================================
// Scanner DTOs
// =========================================================================

// ScanSiteRequest contains parameters for scanning a WordPress site
type ScanSiteRequest struct {
	SiteID        int64  `json:"siteId"`
	SitemapName   string `json:"sitemapName"`
	TitleSource   string `json:"titleSource"`   // "title" or "h1"
	ContentFilter string `json:"contentFilter"` // "all", "pages", or "posts"
	IncludeDrafts bool   `json:"includeDrafts"`
	MaxDepth      int    `json:"maxDepth"` // 0 = unlimited
}

// ScanSiteResponse contains the result of a site scan
type ScanSiteResponse struct {
	SitemapID     int64       `json:"sitemapId"`
	PagesScanned  int         `json:"pagesScanned"`
	PostsScanned  int         `json:"postsScanned"`
	NodesCreated  int         `json:"nodesCreated"`
	NodesSkipped  int         `json:"nodesSkipped"`
	TotalDuration string      `json:"totalDuration"`
	Errors        []ScanError `json:"errors,omitempty"`
}

// ScanError represents an error during scanning
type ScanError struct {
	WPID    int    `json:"wpId,omitempty"`
	Type    string `json:"type,omitempty"` // "page" or "post"
	Title   string `json:"title,omitempty"`
	Message string `json:"message"`
}

// ScanIntoSitemapRequest contains parameters for scanning into an existing sitemap
type ScanIntoSitemapRequest struct {
	SitemapID     int64  `json:"sitemapId"`
	ParentNodeID  *int64 `json:"parentNodeId,omitempty"` // Optional: if nil, uses root node
	TitleSource   string `json:"titleSource"`            // "title" or "h1"
	ContentFilter string `json:"contentFilter"`          // "all", "pages", or "posts"
	IncludeDrafts bool   `json:"includeDrafts"`
	MaxDepth      int    `json:"maxDepth"` // 0 = unlimited
}

// =========================================================================
// Sync DTOs
// =========================================================================

// SyncNodesRequest contains parameters for syncing nodes from WordPress
type SyncNodesRequest struct {
	SiteID  int64   `json:"siteId"`
	NodeIDs []int64 `json:"nodeIds"`
}

// SyncNodesResponse contains the result of syncing nodes
type SyncNodesResponse struct {
	Results []SyncNodeResult `json:"results"`
}

// SyncNodeResult contains the result of syncing a single node
type SyncNodeResult struct {
	NodeID  int64  `json:"nodeId"`
	Success bool   `json:"success"`
	Error   string `json:"error,omitempty"`
}

// UpdateNodesToWPRequest contains parameters for updating WordPress from local nodes
type UpdateNodesToWPRequest struct {
	SiteID  int64   `json:"siteId"`
	NodeIDs []int64 `json:"nodeIds"`
}

// ChangePublishStatusRequest contains parameters for changing publish status
type ChangePublishStatusRequest struct {
	SiteID    int64  `json:"siteId"`
	NodeID    int64  `json:"nodeId"`
	NewStatus string `json:"newStatus"` // "published", "draft", "pending"
}

// =========================================================================
// AI Generation DTOs
// =========================================================================

// TitleInput represents a single title with optional keywords for AI generation
type TitleInput struct {
	Title    string   `json:"title"`
	Keywords []string `json:"keywords,omitempty"`
}

// GenerateSitemapStructureRequest contains parameters for AI sitemap generation
type GenerateSitemapStructureRequest struct {
	// SitemapID is set when adding to existing sitemap, nil for new sitemap
	SitemapID *int64 `json:"sitemapId,omitempty"`
	// SiteID is required when creating new sitemap
	SiteID *int64 `json:"siteId,omitempty"`
	// Name for new sitemap (required when SitemapID is nil)
	Name string `json:"name,omitempty"`
	// PromptID is the ID of the prompt template to use
	PromptID int64 `json:"promptId"`
	// Placeholders for the prompt template
	Placeholders map[string]string `json:"placeholders,omitempty"`
	// Titles with optional keywords to generate structure for
	Titles []TitleInput `json:"titles"`
	// ParentNodeIDs - nodes to use as roots for new nodes (empty = use root node)
	ParentNodeIDs []int64 `json:"parentNodeIds,omitempty"`
	// MaxDepth limits the depth of generated structure (0 = no limit)
	MaxDepth int `json:"maxDepth,omitempty"`
	// IncludeExistingTree sends the current tree structure to AI for context
	IncludeExistingTree bool `json:"includeExistingTree,omitempty"`
	// ProviderID is the AI provider to use
	ProviderID int64 `json:"providerId"`
}

// GeneratedNode represents a single node in AI-generated structure
type GeneratedNode struct {
	Title    string          `json:"title"`
	Slug     string          `json:"slug"`
	Keywords []string        `json:"keywords,omitempty"`
	Children []GeneratedNode `json:"children,omitempty"`
}

// GenerateSitemapStructureResponse contains the AI-generated sitemap structure
type GenerateSitemapStructureResponse struct {
	SitemapID    int64 `json:"sitemapId"`
	NodesCreated int   `json:"nodesCreated"`
	DurationMs   int64 `json:"durationMs"`
}

// ExistingNodeInfo represents simplified node info sent to AI for context
type ExistingNodeInfo struct {
	Title    string             `json:"title"`
	Slug     string             `json:"slug"`
	Path     string             `json:"path"`
	Keywords []string           `json:"keywords,omitempty"`
	Children []ExistingNodeInfo `json:"children,omitempty"`
}

// =========================================================================
// History DTOs
// =========================================================================

// HistoryState represents the current state of the history stack
type HistoryState struct {
	CanUndo       bool   `json:"canUndo"`
	CanRedo       bool   `json:"canRedo"`
	UndoCount     int    `json:"undoCount"`
	RedoCount     int    `json:"redoCount"`
	LastAction    string `json:"lastAction,omitempty"`
	ActionApplied string `json:"actionApplied,omitempty"` // Description of the action that was just applied (undo/redo)
}

// =========================================================================
// Page Generation DTOs
// =========================================================================

type ContentSettingsDTO struct {
	WordCount               string `json:"wordCount"`                         // e.g. "1000" or "800-1200"
	WritingStyle            string `json:"writingStyle"`                      // professional, casual, formal, friendly, technical
	ContentTone             string `json:"contentTone"`                       // informative, persuasive, educational, engaging, authoritative
	CustomInstructions      string `json:"customInstructions"`                // Additional instructions
	UseWebSearch            bool   `json:"useWebSearch"`                      // Enable web search for AI generation
	IncludeLinks            bool   `json:"includeLinks"`                      // Include approved links from linking plan
	AutoLinkMode            string `json:"autoLinkMode,omitempty"`            // "none", "before", or "after"
	AutoLinkProviderID      *int64 `json:"autoLinkProviderId,omitempty"`      // Provider for link suggestion
	AutoLinkSuggestPromptID *int64 `json:"autoLinkSuggestPromptId,omitempty"` // Prompt for link suggestion (link_suggest)
	AutoLinkApplyPromptID   *int64 `json:"autoLinkApplyPromptId,omitempty"`   // Prompt for link insertion (link_apply)
	MaxIncomingLinks        int    `json:"maxIncomingLinks,omitempty"`        // Max incoming links per page
	MaxOutgoingLinks        int    `json:"maxOutgoingLinks,omitempty"`        // Max outgoing links per page
}

// LinkTargetDTO represents a target page for internal linking during generation
type LinkTargetDTO struct {
	TargetNodeID int64   `json:"targetNodeId"`
	TargetTitle  string  `json:"targetTitle"`
	TargetPath   string  `json:"targetPath"`
	AnchorText   *string `json:"anchorText,omitempty"` // Suggested anchor text (optional)
}

type StartPageGenerationRequest struct {
	SitemapID       int64               `json:"sitemapId"`
	NodeIDs         []int64             `json:"nodeIds,omitempty"`
	ProviderID      int64               `json:"providerId"`
	PromptID        *int64              `json:"promptId,omitempty"`
	PublishAs       string              `json:"publishAs"`
	Placeholders    map[string]string   `json:"placeholders,omitempty"`
	MaxConcurrency  int                 `json:"maxConcurrency,omitempty"`
	ContentSettings *ContentSettingsDTO `json:"contentSettings,omitempty"`
}

type GenerationTaskResponse struct {
	ID             string               `json:"id"`
	SitemapID      int64                `json:"sitemapId"`
	SiteID         int64                `json:"siteId"`
	TotalNodes     int                  `json:"totalNodes"`
	ProcessedNodes int                  `json:"processedNodes"`
	FailedNodes    int                  `json:"failedNodes"`
	SkippedNodes   int                  `json:"skippedNodes"`
	Status         string               `json:"status"`
	StartedAt      string               `json:"startedAt"`
	CompletedAt    *string              `json:"completedAt,omitempty"`
	Error          *string              `json:"error,omitempty"`
	Nodes          []GenerationNodeInfo `json:"nodes,omitempty"`
	// Linking phase tracking
	LinkingPhase string `json:"linkingPhase,omitempty"` // "none", "suggesting", "applying", "completed"
	LinksCreated int    `json:"linksCreated,omitempty"` // Number of links suggested
	LinksApplied int    `json:"linksApplied,omitempty"` // Number of links applied
	LinksFailed  int    `json:"linksFailed,omitempty"`  // Number of links that failed to apply
}

type GenerationNodeInfo struct {
	NodeID      int64   `json:"nodeId"`
	Title       string  `json:"title"`
	Path        string  `json:"path"`
	Status      string  `json:"status"`
	ArticleID   *int64  `json:"articleId,omitempty"`
	WPPageID    *int    `json:"wpPageId,omitempty"`
	WPURL       *string `json:"wpUrl,omitempty"`
	Error       *string `json:"error,omitempty"`
	StartedAt   *string `json:"startedAt,omitempty"`
	CompletedAt *string `json:"completedAt,omitempty"`
}

type DefaultPromptResponse struct {
	Name         string   `json:"name"`
	SystemPrompt string   `json:"systemPrompt"`
	UserPrompt   string   `json:"userPrompt"`
	Placeholders []string `json:"placeholders"`
}
